src/conn/conn_open.c

```diff
@@ -100,6 +94,10 @@ __wt_connection_close(WT_CONNECTION_IMPL *conn)
 		__wt_yield();
 	}
 
+	/* Shut down the subsystems, ensuring workers see the state change. */
+	F_SET(conn, WT_CONN_CLOSING);
```

Comment:
@agorrod 
> I believe this was after the subsystem destroy calls so that those subsystems had a chance to do an orderly shutdown of their worker threads. I'm not sure if that's currently required - but I think it's safer to keep the ordering.

It's currently required because there's an assert when opening a Btree that we're not in the "closing" path, and that's why we didn't set `WT_CONN_CLOSING` until after async and LSM had exited.

> Why the order change in here?

The order change is because I wanted `WT_CONN_CLOSING` to mean "closing", rather than "at the point in closing where we no longer should open files". I've added a new flag `WT_CONN_CLOSING_NO_MORE_OPENS` to support the previous diagnostic check, hopefully that works for everybody!

> Why is the barrier required? Do some operations need to see the change to WT_CONN_CLOSING?

I doubt the barrier is required (I did the same review as you, and I agree no thread needs to see the flag at this time). That said: it's a flag other threads are expected to use to modify their actions, and it should have a barrier for that reason. Asking a future programmer to realize `WT_CONN_CLOSING` shouldn't be used for anything critical because it just might not be visible, is an unreasonable requirement.